ENH: Discuss adding anchor links to sidebar

rossbar opened the pull request:
This PR demonstrates changes that could be made to the `sidebar.html` liquid template to add anchoring to links in the sidebar.

I'm adding the PR to illustrate how I understood the feature suggestion in the top comment of #191.

See issue #191 itself for more detailed discussion

choldgraf commented:
Thanks for scoping this out! I think this is close to what we'd want for within-page anchors in the sidebar. A couple of things to ensure:

* The main thing that I'm not sure about is how titles would get rendered. E.g., if somebody had a page w/ 2 anchors under it, each linked in the sidebar, then would each one have <pagetitle> as its title?
* How would this affect the HTML generation process? Right now we use the `toc.yml` file to decide which content files to generate into HTML. There's an assumption of "one toc.yml entry per content file". This would break that assumption if you have multiple toc entries for sub-sections in a file, so we should make sure this wouldn't create unexpected behavior in the builds.

rossbar commented:
You've definitely hit on important points:

> The main thing that I'm not sure about is how titles would get rendered. E.g., if somebody
> had a page w/ 2 anchors under it, each linked in the sidebar, then would each one have as its title?

As of commit 8984035, unless the `title:` keyword is used, the title will be the same as the parent page. As an example of what would happen, here is a modified `toc.yaml` that creates anchor links for all of the section headings in the FAQ:

```yaml
 - url: /guide/05_faq
    expand_subsections: true
    subsections:
    - url: /guide/05_faq
      anchor: How-can-I-update-my-book?
    - url: /guide/05_faq
      anchor: Does-the-book-behave-differently-depending-on-the-browser?
    - url: /guide/05_faq
      anchor: Why-isn't-my-math-showing-up-properly?
    - url: /guide/05_faq
      anchor: How-can-I-include-interactive-Plotly-figures?
    - url: /guide/05_faq
      anchor: What-if-I-have-an-issue-or-question?
    - url: /guide/05_faq
      anchor: How-should-I-add-cell-tags-to-my-notebooks?
    - url: /guide/05_faq
      anchor: Why-doesn't-math-show-properly-when-printing?
    - url: /guide/05_faq
      anchor: What-is-this-Gemfile.lock-file-or-why-am-I-getting-Jekyll-dependency-warnings?
```
And the resulting sidebar:

![image](https://user-images.githubusercontent.com/1268991/70867378-93157a80-1f29-11ea-8252-c2e3af3309ba.png)

This is obviously not the desired behavior. This can be fixed by manually specifying the `title:` tag to go along with each anchor, but that seems tedious as well. I was considering adding machinery to infer the title from the `anchor:` link.

> How would this affect the HTML generation process? Right now we use the toc.yml file to
> decide which content files to generate into HTML. There's an assumption of "one toc.yml
> entry per content file". This would break that assumption if you have multiple toc entries for
> sub-sections in a file, so we should make sure this wouldn't create unexpected behavior in
> the builds.

The current solution definitely does have an unintended side-effect for the build process: it causes a build of the page for each anchor. For example, when building the above example, `/guide/05_faq` is built 9 times (once for the page itself, plus once for each extra anchor keyword). This obviously needs to be fixed as well. 

I plan to keep working on it!

rossbar commented:
Okay, I think I've hit upon a solution that addresses the above issues. I decided to see how far I could get by leaving the build machinery alone and relying on the liquid templates to do all of the work for internal linking:
  1. I've modified `_prepare_toc` to automatically remove toc entries that have an anchor defined.
      This keeps the "one file per toc entry" assumption valid downstream in the build process (and 
      solves the problem a single file being built multiple times).
  2. I've added additional logic to the `sidebar.html` liquid template to check for anchors. Basically, 
      if liquid finds an `anchor` in a toc entry, it assigns the sidebar title first by checking to see if a
      `title:` has been explicitly specified, then falling back on the anchor name itself if not. 

Going back to the FAQ example, here's how things look now:

```yaml
  - url: /guide/05_faq                                                          
    expand_subsections: true                                                    
    subsections:                                                                
    - url: /guide/05_faq                                                        
      title: Updating a Book                                                    
      anchor: How-can-I-update-my-book?                                         
    - url: /guide/05_faq                                                        
      title: Browser Differences                                                
      anchor: Does-the-book-behave-differently-depending-on-the-browser?        
    - url: /guide/05_faq                                                        
      title: Math Rendering                                                     
      anchor: Why-isn't-my-math-showing-up-properly?                            
    - url: /guide/05_faq                                                        
      title: Plotly                                                             
      anchor: How-can-I-include-interactive-Plotly-figures?                     
    - url: /guide/05_faq                                                        
      anchor: What-if-I-have-an-issue-or-question?                              
    - url: /guide/05_faq                                                        
      anchor: How-should-I-add-cell-tags-to-my-notebooks?                       
    - url: /guide/05_faq                                                        
      anchor: Why-doesn't-math-show-properly-when-printing?                     
    - url: /guide/05_faq                                                        
      anchor: What-is-this-Gemfile.lock-file-or-why-am-I-getting-Jekyll-dependency-warnings?
```
Which results in a sidebar like this (see https://rossbar.github.io/jupyterbook-site-test/intro):

![image](https://user-images.githubusercontent.com/1268991/70878738-84ee4b00-1f77-11ea-8e78-71f5fc00ccd4.png)

Let me know if you see any problems with this approach. If not, I'm happy to write some tests & docs and submit for review.

rossbar commented on jupyter_book/book_template/_includes/sidebar.html:
```diff
@@ -35,17 +35,29 @@ <h2 class="c-sidebar__title">{{ site.title }}</h2>
+        {% if chapter.title %}
+          {% assign chapter_title = chapter.title %}
+        {% else %}
+          {% assign chapter_title = chapter.anchor | replace: "-", " " %}
```
I'm not sure - I think the names for the anchor links are determined by `anchor.js`, which I'm not familiar with. I don't know if the HTML-ifying pattern that `anchor.js` uses is the default or a standard or whether it will change in the future.

rossbar commented:
Thanks for taking a look!

 > Does this mean that if somebody only links to a page using an anchor link, then it won't be built?

Yes, that is the current behavior. Since this PR deals with a way to add anchor links to the sidebar only, I had been operating on the assumption that the anchor links would only be generated as sub-elements of an already-built page. I hadn't considered the case where the user would want to *only* refer to sub-headers in a single page without ever explicitly referring to page itself in the toc.

choldgraf commented on jupyter_book/build.py:
```diff
@@ -51,10 +51,11 @@ def _prepare_toc(toc):
     return [
         item for item in new_toc
-        if 'url' in item and not item.get('external', False)
+        if 'url' in item and not item.get('external', False) and item.get('anchor') is None
```
here rather than removing anything that has an anchor, why don't we instead just remove any redundant items from this returned list. So something like

```python
all_pages = [
        item for item in new_toc
        if 'url' in item and not item.get('external', False)
]
used_urls = []
pages_to_build = []
for page in all_pages:
    if page.url in used_urls:
        pass
    pages_to_build.append(page)
    used_urls.append(page.url)
return pages_to_build
```

I think that'd ensure that we don't double-build pages, and would also include pages if they *only* showed up w/ an anchor link

rossbar replied:
I agree, the mechanism you describe removes the limitation that anchor links can only be subordinate to already-built pages. I think I was trying to be too cute: I had it as a goal not to clutter up the code too much because it was so easy to read!

It's too bad the `items` are dicts because if they were hashable I think the same behavior could be achieved by replacing the `all_pages` list comp with a set comp.

rossbar replied:
I've incorporated your suggested change from this discussion. 

There are still a few outstanding issues:

 1. The anchor links don't work when they have apostrophe's in them (e.g. try clicking on the "Math Rendering" sidebar item on [my build of the jupyter-book website](https://rossbar.github.io/jupyterbook-site-test/intro.html)). I thought this might be a simple html-encoding issue, but unfortunately it seems like the `url_encode` liquid filter and anchorjs do not play nice with each other. I will look for a different solution.

 2. There is an annoying behavior in the sidebar: if you click a top-level link in the sidebar that has anchors, all of the anchor links light up blue (e.g. click on the "Frequently Asked Questions (FAQ)" link in the sidebar on on [my build of the jupyter-book website](https://rossbar.github.io/jupyterbook-site-test/intro.html)). I think this has to do with the fact that the `data-url` of the sidebar items in `sidebar.html` refers to the same top-level link even when an anchor is applied. Nothing is necessarily "broken", but it is ugly so I'd like to try to fix it.

One conclusion that I'm coming to by working on this is that I find liquid to be very cumbersome. I have no problem working on this and will continue to do so, but if you think that my approach is opening up too much of a can-of-worms please let me know!

choldgraf replied:
> One conclusion that I'm coming to by working on this is that I find liquid to be very cumbersome.

One of the reasons it's worth exploring Sphinx + Jinja? At least then we could control some of the logic within Python instead of just within the template language. For example, [here the pydata sphinx theme defines its own python function to get the nav object](https://github.com/pandas-dev/pydata-bootstrap-sphinx-theme/blob/master/pandas_sphinx_theme/__init__.py#L64) and then [it loads the nav object into Jinja here](https://github.com/pandas-dev/pydata-bootstrap-sphinx-theme/blob/master/pandas_sphinx_theme/docs-navbar.html#L16)

choldgraf commented:
@rossbar I'm gonna close this one, as this should be superceded by beta.jupyterbook.org, but thanks very much for your work on this! We should explore if and how to accomplish this in the new Sphinx build system

rossbar commented:
Agreed - thanks for the ping!
